code/modules/mob/living/carbon/xenomorph/castes/drone/abilities_drone.dm

```diff
@@ -3,20 +3,107 @@
+		return FALSE
+
+	if(target.stat != DEAD)
+		to_chat(X, "<span class='xenowarning'>We can't salvage the biomass of living sisters!</span>")
+		return FALSE
+
+	var/distance = get_dist(X, target)
+	if(distance > DRONE_SALVAGE_BIOMASS_RANGE)
+		to_chat(X, "<span class='xenowarning'>We need to be [distance -  DRONE_SALVAGE_BIOMASS_RANGE] tiles closer to [target].</span>")
+		return FALSE
+
+
+/datum/action/xeno_action/activable/salvage_biomass/use_ability(atom/A)
+
+	var/mob/living/carbon/xenomorph/X = owner
+	var/mob/living/carbon/xenomorph/target = A
+
+
+	X.face_atom(target) //Face the target so we don't look like an ass
+	to_chat(X, "<span class='xenowarning'>We being to salvage the biomass of [target]...</span>")
+	playsound(X, pick('sound/voice/alien_drool1.ogg','sound/voice/alien_drool2.ogg'), 25)
+
+	if(!do_after(X, DRONE_SALVAGE_BIOMASS_WINDUP, TRUE, target, BUSY_ICON_FRIENDLY)) //Wind up
+		return fail_activate()
+
+
+	X.face_atom(target) //Face the target so we don't look like an ass
+
+	if(!can_use_ability(target, FALSE, XACT_IGNORE_PLASMA)) //Check to make sure the target hasn't gone anywhere/been exploded/etc
+		return fail_activate()
+
+	succeed_activate()
+
+	var/plasma_amount = target.plasma_stored //We take all the target's plasma; they won't need it where they're going/are
+	var/upgrade_amount = target.upgrade_stored * DRONE_SALVAGE_BIOMASS_SALVAGE_RATIO //We only recover a small portion of the target's upgrade and evo points.
+	var/evo_amount = target.evolution_stored * DRONE_SALVAGE_BIOMASS_SALVAGE_RATIO
+
+	//Take all the plas-mar
+	target.use_plasma(plasma_amount)
+	X.gain_plasma(plasma_amount)
+
+	//Distribute the upgrade and evo points the target had to the hive:
+	var/list/list_of_upgrade_xenos = list()
+	var/list/list_of_evolve_xenos = list()
+
+	var/list/prohibited_upgrades = list(XENO_UPGRADE_THREE, XENO_UPGRADE_INVALID)
+	var/list/prohibited_tiers = list(XENO_TIER_ZERO, XENO_TIER_THREE, XENO_TIER_FOUR)
+
+	for(var/mob/living/carbon/xenomorph/filter in X.hive.get_all_xenos())
+		if(!prohibited_upgrades.Find(filter.upgrade)) //Only Xenos who can use the salvage get it; filter them
+			list_of_upgrade_xenos += filter
+		if(!prohibited_tiers.Find(filter.tier))
+			list_of_evolve_xenos += filter
+
+	var/number_of_upgrade_xenos = length(list_of_upgrade_xenos)
+	var/number_of_evolve_xenos = length(list_of_evolve_xenos)
+
+	upgrade_amount /= max(1, number_of_upgrade_xenos) //get the amount distributed to each xeno; protect against dividing by 0
+	evo_amount /= max(1, number_of_evolve_xenos)
+
+	for(var/mob/living/carbon/xenomorph/beneficiary in list_of_upgrade_xenos) //Distribute the upgrade salvage to those who can use it
+		beneficiary.upgrade_stored += upgrade_amount
+
+	for(var/mob/living/carbon/xenomorph/beneficiary in list_of_evolve_xenos) //Distribute the upgrade salvage to those who can use it
+		beneficiary.evolution_stored += evo_amount
```

Comment:
handle this in for(var/mob/living/carbon/xenomorph/filter in X.hive.get_all_xenos())